### ISSUE TITLE
Thread bug with Rails::Engine mounted routes, undefined method `url_options'

### ISSUE THREAD

**Issue by sbull** [rails/rails#19084]
This issue occurs randomly in production, as it's a multi-threading problem. This is very related to https://github.com/puma/puma/issues/647, and the underlying MRI bug https://bugs.ruby-lang.org/issues/10871.

I'm using a mounted engine, rails_admin, and I get the following exception frequently in one of my views, in which I have `rails_admin.show_path(model_name: 'dummy', id: 123)`.

```
ActionView::Template::Error: undefined method `url_options' for #<Module:0x007f432e67dc20>
action_dispatch/routing/route_set.rb:271:in `call'
    controller_options = t.url_options
  action_dispatch/routing/route_set.rb:222:in `call'
    super
  action_dispatch/routing/route_set.rb:334:in `block (2 levels) in define_url_helper'
    helper.call self, args, options
  action_dispatch/routing/mapper.rb:637:in `block (2 levels) in define_generate_prefix'
    _routes.url_helpers.send("#{name}_path", prefix_options)
  action_dispatch/routing/route_set.rb:775:in `url_for'
    script_name = find_script_name options
  action_dispatch/routing/route_set.rb:279:in `call'
    t._routes.url_for(hash, route_name, url_strategy)
  action_dispatch/routing/route_set.rb:222:in `call'
    super
  action_dispatch/routing/route_set.rb:334:in `block (2 levels) in define_url_helper'
    helper.call self, args, options
...
```

I have isolated the problem with a reproducible test script and example output here:
https://gist.github.com/sbull/78eb41fdfd0460f5ce80#file-route_helpers_debug-rb

I am able to work around thread problems with the use of Rails.application.routes.url_helpers in other contexts in my app, but I haven't yet found a way to workaround this problem with routes for mounted engines. Hopefully there's a way for these mounted url helpers to work around the underlying bug in MRI.

/cc @evanphx, /cc @tenderlove (due to involvement in previous issue)

**Comment by evanphx** [comment]
This should be fixed by https://github.com/rails/rails/pull/19007 by not invoking url_helpers concurrently.

**Comment by sbull** [comment]
Would any workarounds be apparent in the meantime?

**Comment by sbull** [comment]
I'm attempting to create a workaround for this using a Mutex. I've been able to greatly reduce the frequency of errors, but not eliminate it completely. Here's the details: https://gist.github.com/sbull/e83defaa7c6a93950866

I'm including the important bits here:
```ruby
class TestEngine < Rails::Engine; end
TestEngine.routes.draw do
  get '/' => 'dummy#index', as: :dummy
end

Rails.application.routes.draw do
  mount TestEngine, at: '-', as: 'test_routes'
end

class TestController < ActionController::Base
  def engine_test
    render plain: RouteHelper.one.dummy_path
  end
end

ROUTES_MUTEX = Mutex.new

class RouteHelper

  class UrlHelpers
    include TestEngine.routes.url_helpers
  end

  def self.one
    @@one
  end

  def self.one=(one)
    @@one = one
  end

  def self.url_helpers=(url_helpers)
    @@url_helpers = url_helpers
  end

  def method_missing(symbol, *args)
    ROUTES_MUTEX.synchronize {
      @@url_helpers.send(symbol, *args)
    }
  end

end

RouteHelper.one = RouteHelper.new
RouteHelper.url_helpers = RouteHelper::UrlHelpers.new

puts "Running Mounted Engine dummy_path..."
run_threads do
  env = { 'REQUEST_METHOD' => 'GET', 'rack.input' => StringIO.new }
  status, headers, rack_body = TestController.action('engine_test').call(env)
  rack_body.body
end
```
Now, as best as I can reason, the ROUTES_MUTEX would prevent any concurrent invoking of the buggy url_helpers. I've tried a few other variations of this, and I still get an error every few times I run the test script.

This makes me concerned that there's something causing problems that I don't understand. As far as I can see, the real url_helpers would never be called concurrently. @evanphx, is there any chance that you can take a peek to let me know where this workaround is failing?

**Comment by evanphx** [comment]
@sbull you could probably take the change I did in #19007 and make it a hot patch you put into your initializers, to just fix the core rails behavior.

**Comment by jvanbaarsen** [comment]
@rafaelfranca can you add the label "With reproduction steps"?

**Comment by sbull** [comment]
@evanphx I've tried adding a monkeypatch with your fix in #19007 to my original reproduction script. It has the result of reducing the frequency of the problem, but not fixing it. (Originally I was getting an average of 8 failed threads out of 10, with the monkeypatch I get an average of 5 failed threads out of 10.) Here's the reproduction with the monkeypatch: https://gist.github.com/sbull/5d716e74bcc323b29895

When I tried my workaround approach above with a Mutex, I would get either no failures or 1 failure, regardless of how many threads or iterations per thread I tried.

**Comment by sbull** [comment]
If anyone is interested in an "almost fixed" workaround, here's what I've done:
```ruby
module Routes
  class RailsAdmin
    class UrlHelpers
      include ::RailsAdmin::Engine.routes.url_helpers
    end
    @@semaphore = Mutex.new
    @@url_helpers = UrlHelpers.new
    def method_missing(symbol, *args)
      @@semaphore.synchronize {
        @@url_helpers.send(symbol, *args)
      }
    end
    def self.solo
      @@solo
    end
    @@solo = self.new
  end
  def self.rails_admin
    RailsAdmin.solo
  end
end
```
And I use it like `Routes.rails_admin.show_path(model_name: 'post', id: 123)`. So far, this works in real life, but produces an occasional error in my contrived reproduction script.

**Comment by rafaelfranca** [comment]
@sbull have you tested with an application pointing to master branch?

**Comment by sbull** [comment]
Short answer: problem still exists on master.

I've tried this again using a clone of rails master (`git clone git://github.com/rails/rails.git`, `cd rails`, `bundle install`, `bundle exec rails new ../rails-routes-bug --dev`), and run my test script again (`bundle exec rails runner route_helpers_debug.rb`). It appears that the frequency of the issue is greatly reduced, but it still happens occasionally. In fact, I've had to increase the number of threads/iterations to exhibit the problem.

Here's the updated gist with the script and the output: https://gist.github.com/sbull/114f93a374136ace74f1

**Comment by rafaelfranca** [comment]
Thans

**Comment by ahacop** [comment]
@sbull Is this still a problem? Following the steps outlined here I could not reproduce the errors against master.

**Comment by sbull** [comment]
Still a problem. I've just followed the same steps as above (https://github.com/rails/rails/issues/19084#issuecomment-91130754), and run my test script with increased values like `bundle exec rails runner route_helpers_debug.rb 20 30` to make the same errors happen (<code>Exception in thread 17, iteration 28: undefined method `url_options' for #&lt;Module:0x007f7e34e8c808&gt;</code>).

**Comment by evanphx** [comment]
The bug is actually in ruby itself. I believe it was fixed in 2.2 release. /cc @_ko1 @tenderlove

**Comment by ahacop** [comment]
I can confirm that I followed the reproduction steps with the increased values suggested by @sbull and did not get any errors while using 2.2.2.

**Comment by sbull** [comment]
Yeah, I just ran this using 2.2.2 and it still has the bug. Per https://bugs.ruby-lang.org/issues/10871#note-12, I don't think it's merged yet.

@ahacop, can you increase those values and try running several times? Basically, keep increasing them, and eventually you'll run into the problem (it's non-deterministic, probably depends on cpu resources etc.)

**Comment by ahacop** [comment]
@sbull I did eventually run into the error. But, if this is a bug in ruby, can this issue be closed?

**Comment by sbull** [comment]
Well, it's a bug in ruby, but that bug is only exhibited by certain styles of code. So if this Rails feature were implemented differently, it may not have the bug.

**Comment by evanphx** [comment]
@ahacop It's up to the rails contributors, but I'd leave it open because people running rails hit it and wonder where it's coming from.

**Comment by dlackty** [comment]
MRI Ruby 2.2.4 was released and it should fix the issue. 

https://www.ruby-lang.org/en/news/2015/12/16/ruby-2-2-4-released/

**Comment by morgoth** [comment]
I confirm that error doesn't happen anymore on Ruby 2.3

**Comment by guyisra** [comment]
is there a planned fixed for this when using Ruby 2.1.x and without monkeypatching?

**Comment by skipkayhil** [comment]
It sounds like this has been fixed in all supported Rubies so I believe this can be closed now.
